language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -242,34 +259,34 @@ computeCek con t@(Var _ varName)   = do
 returnCek [] res = getVarEnv <&> \varEnv -> void $ dischargeVarEnv varEnv res
-returnCek (FrameTyInstArg ty : con) fun = instantiateEvaluate con ty fun
-returnCek (FrameApplyArg argVarEnv arg : con) fun = do
+returnCek (FrameTyInstArg ty : ctx) fun =
+    case fun of
+      TyAbs _ _ _ body -> computeCek ctx body  -- NB: we're just ignoring the type here
+      _                ->  throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just (void fun)
+returnCek (FrameApplyArg argVarEnv arg : ctx) fun = do
     funVarEnv <- getVarEnv
-    withVarEnv argVarEnv $ computeCek (FrameApplyFun funVarEnv fun : con) arg
-returnCek (FrameApplyFun funVarEnv fun : con) arg = do
-    argVarEnv <- getVarEnv
-    applyEvaluate funVarEnv argVarEnv con fun arg
-returnCek (FrameIWrap ann pat arg : con) val =
-    returnCek con $ IWrap ann pat arg val
-returnCek (FrameUnwrap : con) dat = case dat of
-    IWrap _ _ _ term -> returnCek con term
+    withVarEnv argVarEnv $ computeCek (FrameApplyFun funVarEnv fun : ctx) arg
+returnCek (FrameApplyFun funVarEnv fun : ctx) arg =
+    case fun of
+      LamAbs _ name _ body -> do
+               argVarEnv <- getVarEnv
+               withVarEnv (extendVarEnv name arg argVarEnv funVarEnv) $ computeCek ctx body
+      _ ->  throwingWithCause _MachineError NonPrimitiveApplicationMachineError $ Just (Apply () (void fun) (void arg))
+      -- FIXME: ^ error message looks like eg (1 2) instead of [(con 1) (con 2)]
+returnCek (FrameIWrap ann pat arg : ctx) val =
+    returnCek ctx $ IWrap ann pat arg val
+returnCek (FrameUnwrap : ctx) dat = case dat of
+    IWrap _ _ _ term -> returnCek ctx term
     term             ->
         throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just (void term)
-
--- | Instantiate a term with a type and proceed.
--- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
--- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
--- apply the term to the type via 'TyInst'.
-instantiateEvaluate
-    :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => Context uni -> Type TyName uni ExMemory -> WithMemory Term uni -> CekM uni (Plain Term uni)
-instantiateEvaluate con _ (TyAbs _ _ _ body) = computeCek con body
-instantiateEvaluate con ty fun
-    | isJust $ termAsPrimIterApp fun = returnCek con $ TyInst (memoryUsage () <> memoryUsage fun <> memoryUsage ty) fun ty
-    | otherwise                      =
-        throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just (void fun)
+returnCek (FrameApplyBuiltin env bn tys closures terms : ctx) value = do
+    varEnv <- getVarEnv
+    let closures' = Closure varEnv value : closures -- Save the closure for the argument we've just computed.
+    case terms of
+      []       -> applyBuiltin env () ctx bn tys (reverse closures') -- We've accumulated the argument closures in reverse
```

Comment:
Yes, but that's a bit fragile because quite a lot goes on between the two reversals, and the fact that everything ends up in the right order is a lucky consequence of the fact that you call two things that reverse their arguments.  I've got into the habit of reversing accumulators at the first possible opportunity because if you forget to do it you can end up with confusing bugs.

Anyway, this will go away if we just merge the functionality of `getScopedArgs` into the frames.